README.md
# Pepe Bot

Pepe bot is a bot I am making to practice Python, and to have a second go at making a Discord bot. 

Warning: Clone this repo at your own discretion, the code may break, fail, cause damage, etc. I have no clue what I'm doing.

# Commands

**To view the commands the bot has, [click here](docs/commands.md)**

# Acknowledgements

## Henrik Dev
- The API by [Henrik Dev](https://github.com/Henrik-3/) was instrumental in the VALORANT part of this bot.

### semaJ
- For all the [agent images](./imgs/agents). 


docs/changelog.md
# PepeBot Changelog

## Aug 11 2021

### New ``=checkapi`` command

- Check the endpoints for responses!
    - See if the API is down, as well as which endpoints are affected!

### ``=checkapi`` moved to ``=apistats``

- The API stats have moved to ``=apistats``
- New ``=checkapi`` command

## Aug 10 2021

### ``=peak`` command!

- Check someones peak rank!
    - Shows the act they peaked
    - Shows how long they peaked for
    - Shows their current rank

### Testing new commands

- Will be released to beta soon.

## Jul 26 2021

### ``=mmr`` moved to Public Access

- The command is now open source under the ``mmrhistory.py`` file
- Now can see net gain/loss RR from comp careers

## Jul 22 2021

### Beta command ``=mmr``

- Calculate stats based on all comp games currently viewable in specified players career
- Command is in beta and won't work most of the time
- Will be moved out of beta shortly

## Jul 14 2021

### Updated ``=checkapi`` command!

- Added check of latest request

### Return of ``=valm`` command!

- Use ``=valm USERNAME#TAG`` to get the last match stats of a player
- Cooldown of 16 seconds

## Jul 13 2021

### Updated ``=val`` command!

- Get up to 5 players ranks at once!
- Split the players with ``\\`` between each player
- Max of FIVE players
- Circumvent the 16 second cooldown using this new way.

# IF PLAYER HAS ``//`` IN NAME:

- Use '\' key to "escape" the ``/``.
- Ex:
    ``=val thelegend\//27#1234``
- (you cannot use multiple people if one of them has ``//`` in their name)

## Jul 8 2021

### New ``=level`` command!

- Check people's VALORANT level from Discord!
#### I'm well aware of a bug causing it to appear uncentered, I cannot do much for it unfortunately.

### New ``=checkapi`` command!

- See how the api has been performing on PepeBot!
- The command shows error counts, error codes, and success rates.
- It also shows a start date from which caching began.

## Jul 7 2021

### New ``=val`` command!

- ``=val`` v2 is out!
    - New fresh color
    - Support for Account Level
    - Redesigned embed
    - Disclaimer in footer
- ``career`` is much stricter access
- ``career`` now clears first 2 messages on success (less spammy)
- Reduced cooldown from 30 >> 16 seconds

## Jul 5 2021

### Patches

- Patched a bug with starting the bot
- Removed commands ``=talk`` and ``=search``

### NOTICE: 

**I had to rollback this changelog so it is slightly inaccurate and missing data from Jun 14-Jul 5, sorry.**
**Stable Release soon**

## Jun 14 2021

### Bug Fixes

- Fixed a ``=vala`` bug where Unrated games would show as Unknown

## Jun 11 2021

### New Command

- Added a rule command (rr) ~~you might want to test it~~

### Jun 9 2021

### Major Bug Fixes

- Fixed bug with ``=lb``
- Rewrote the ``=lb`` command, so it works well now

### Bug Fixes

- Fixed a bug where ``=changelog cb`` command would be over 2000 chars, now should send 1990, then ``....``

### Report command

- Report bugs to the dev with the ``=report BUG HERE`` command!

### Jun 7 2021

### Major Bug Fixes

- Issued a much more permanent fix for ``=vala`` and ``=career`` commands
- Issued a bug fix for certain other commands

- VAL COMMAND may not work!
    - This is an API ISSUE.
        - There is nothing I can do to fix this, as the API is the one with the problem, not me.


### Jun 6 2021

### Lots of new features

- Added features for different commands
- Various bug fixes
- WIP bug fix for vala and career, might be an api issue
- WIP bug fix for github push, should be simple fix

### May 20 2021

### Updates to Leaderboard auth-ed command

- The leaderboard command has always been authorized users only.
- You can now get the leaderboard of any of your past 5 games! 
    - To do this, you will now have to use command ``=lb PLAYER#TAG#GAMENUMBER`` 
    <sub>(Note: Game number must be between 1-5)</sub>
- This continues in this update, but to authorized users, the leaderboard now is much cleaner.
- The leaderboard colors have been uplifted as well.
- Fixed bug where everyone had the same amount of kills and deaths.
- Started working on auth command (wow!)

### New Command

**TEMPERATURE COMMAND** 

- Added Fareinheit to Celcius coverter 
    - ``=tempfc NUM``

- Still in testing, will add to [documentation](./commands.md) later

### May 18 2021

### Bug Fixes

- Fixed a bug where the combat score would divide by the first match round number rather than the current one (Deathmatch games were most affected.)
- Fixed a bug where leaderboards showed:
    - One player only
    - One team only
    - Spammed "` ` `"
    - Spammed "','"
    - Did not return anything
    - Had [] wrapped around the players
    - Had players prefixed with "#"
- Fixed bug where Error commands returned different information
- Fixed by where Error commands contained command names (this was intentional but now is much smoother)
- Fixed bugs with vala
- Fixed bugs where bot would return errors on valm commands
- Fixed bug where wrong errors would return on valm commands

## Release 0.1

**Release 0.1 notes can be found [here](https://github.com/5late/Pepe-Bot/releases/tag/v0.1)**


test.py
from os import path

if path.exists('main.py') and path.exists('github.py'):
    print('The two main files are found. Linting complete.')
else:
    exit(code=1)

script.py
import random

# Code playground for quiz command on bot.
correct_number = False
print(
    "*****Halftime break! You get 3 tries to guess the correct number for an extra point*****\n Rules:\n 1.The number range is 1-5\n 2.The number changes every time you guess\n"
)

for i in range(2):
    total = 0
    while True:
        try:
            guess = int(input("What is your guess? "))
            break
        except ValueError:
            print("Please enter a number!")
            continue
    if guess == random.choice([1, 2, 3, 4, 5]):
        print("\nCONGRATS you get an extra point. The quiz will continue.\n")
        correct_number = True
        total += 1
        break
    else:
        print("Try again\n")

# Last attempt and different statements
if not correct_number:
    while True:
        try:
            guess = int(input("What is your guess? "))
            break
        except ValueError:
            print("Please enter a number!")
            continue
    if guess == random.choice([1, 2, 3, 4, 5]):
        print("CONGRATS you get an extra point. The quiz will continue.\n")
        total += 1
    else:
        print("Good try, the quiz will resume.\n")


careerF.py
import discord
import requests
from discord.ext import commands
from datetime import datetime
import asyncio


async def career(ctx, arg):
    auth = [
        564466359107321856,
        323269361232248832,
        564562239739396098,
        564584121582747659]
    if ctx.author.id not in auth:
        await ctx.send('You are not authorized to use this command. Contact ``Xurxx#7879`` to see how to become authorized.')
    else:
        await ctx.send('You are an authorized user. Results processing...')
        try:
            msg = await ctx.send(f'Your message has been registered and I\'m working hard to query, read, format, and send you the last match for ``{arg}``. Hang tight!')

            newArg = arg.split('#')
            name = newArg[0]
            tag = newArg[1]

            async with ctx.typing():
                response = requests.get(
                    f'https://api.henrikdev.xyz/valorant/v3/matches/na/{name}/{tag}')
                jsonR = response.json()

                embedM = discord.Embed(title=f"{name}\'s Career:")

                for j in range(5):
                    players = jsonR['data']['matches'][j]['players']['all_players']

                    def mode():
                        try:
                            if jsonR['data']['matches'][j]['metadata']['mode'] == 'Normal':
                                return 'Unrated'
                            else:
                                return jsonR['data']['matches'][j]['metadata']['mode']
                        except BaseException:
                            if KeyError:
                                return 'Unknown'

                    def map():
                        try:
                            return str(
                                jsonR['data']['matches'][j]['metadata']['map'])
                        except KeyError:
                            return 'https://upload.wikimedia.org/wikipedia/commons/thumb/4/46/Question_mark_%28black%29.svg/200px-Question_mark_%28black%29.svg.png'

                    for i in players:
                        if i['name'] == name or i['name'] == name.title():
                            team = str(i['team']).lower()
                            if str(mode()) == 'Deathmatch':
                                won = i['stats']['kills']
                                lost = i['stats']['deaths']

                            else:
                                won = jsonR['data']['matches'][j]['teams'][team]['rounds_won']
                                lost = jsonR['data']['matches'][j]['teams'][team]['rounds_lost']

                            embedM.add_field(
                                name='--------------------------',
                                value=f"**{mode()} | {map()}** | ***{won}-{lost}***",
                                inline=False)
                            embedM.add_field(
                                name='Character: ', value=i['character'])
                            embedM.add_field(name='KDA: ', value=str(i['stats']['kills']) + '/' + str(
                                i['stats']['deaths']) + '/' + str(i['stats']['assists']), inline=True)
                            embedM.add_field(
                                name='Combat Score: ',
                                value=int(
                                    i['stats']['score']) // int(
                                    jsonR['data']['matches'][0]['metadata']['rounds_played']),
                                inline=True)
                embedM.description = f'Map/Mode/Score'
                embedM.color = 0xd9d9d9
                if mode() == 'Unknown':
                    embedM.set_footer(
                        text='Wondering what happened to mode/map? Run command ``=error 1``.')
                else:
                    embedM.set_footer(text='https://github.com/5late/Pepe-Bot')
                await asyncio.sleep(1)
                await ctx.send(embed=embedM)
                await msg.edit(content=f':smile: I successfully got the last 5 matches for {name}#{tag}! Congrats on being an authorized user.')
        except BaseException:
            await msg.edit(content='Error 2||Error 404 :(')
            await ctx.send('Use command ``=error 2`` to see more information.')


mmrhistory.py
import discord
import requests
from datetime import datetime


def checkStatusCode(status: str, task: str):
    f = open('./logs/API.txt', 'a')
    now = datetime.now()
    formatTime = now.strftime("%a, %B %d, %Y | %H:%M")
    if status == "200":
        f.write(f'\nSuccesful query on task {task} at {formatTime}.')
    else:
        f.write(f'\nTask {task} returned error code {status} at {formatTime}.')
    f.close()


async def mmrhistory(ctx, arg):
    newArg = arg.split('#')
    name = newArg[0]
    tag = newArg[1]

    response = requests.get(
        f'https://api.henrikdev.xyz/valorant/v1/mmr-history/na/{name}/{tag}')
    jsonR = response.json()

    status = jsonR['status']
    checkStatusCode(status, 'GET-ACCOUNT-MMR-HISTORY')
    if not status == "200":
        return await ctx.send(f'Error occured - status code {status}. Case recorded.')

    games = jsonR['data']
    number_of_games = len(games)

    wins = 0
    losses = 0
    elo = 0
    for game in games:
        if game['mmr_change_to_last_game'] < 0:
            losses += 1
        else:
            wins += 1
        elo += game['mmr_change_to_last_game']

    if wins > losses:
        color = 0x1bfa5e
    elif wins < losses:
        color = 0xff2626
    else:
        color = 0x424744

    if elo > 0:
        elo = f'+{elo}'
    else:
        elo = elo

    latest_game = games[0]['date']

    current_rank = games[0]['currenttier']
    icon_file = discord.File(
        f'./imgs/icons/TX_CompetitiveTier_Large_{current_rank}.png',
        filename='icon.png')

    embed = discord.Embed(
        title=f'MMR History for {name}#{tag}',
        description=f'**{name}** has {number_of_games} competitive games in their career.',
        color=color)
    embed.add_field(name='Wins', value=f'``{wins}``')
    embed.add_field(name='Losses', value=f'``{losses}``')
    embed.add_field(name='Net RR Gained/Lost', value=f'``{elo}``')
    embed.add_field(name='Latest Competitive Game', value=latest_game)
    embed.set_thumbnail(url='attachment://icon.png')

    await ctx.send(file=icon_file, embed=embed)


leaderboard.py
from PIL.Image import new
import discord
from PIL import Image, ImageFont, ImageDraw
import requests
from discord.ext import commands
import asyncio


async def leaderBoard(ctx, arg):
    msg = await ctx.send(f'The command hand-off was successful, I am working hard to sort last game leaderboard information for {arg}.\nPlease allow up to 15 seconds.')

    def lastLetter(word):
        return word[::-1]

    newArg = arg.split('#')
    name = newArg[0]
    tag = newArg[1]

    if len(newArg) == 3:
        lbnum = int(newArg[2]) - 1
    else:
        lbnum = 0

    response = requests.get(
        f'https://api.henrikdev.xyz/valorant/v3/matches/na/{name}/{tag}')
    jsonR = response.json()

    Match = jsonR['data']['matches'][lbnum]
    metadata = Match['metadata']
    players = Match['players']['all_players']
    if metadata['mode'] == 'Deathmatch':
        await msg.edit(content='Error 5 || Error 400')
        await ctx.send('Deathmatch games are not supported.')
    for j in players:
        if j['name'] == name or j['name'] == name.title():
            global team
            team = j['team']
    teamScore = []
    teamSort = []
    enemyScore = []
    enemySort = []

    if team.lower() == 'blue':
        enemyTeam = 'red'
    else:
        enemyTeam = 'blue'

    if Match['teams'][team.lower(
    )]['has_won'] and not Match['teams'][enemyTeam]['has_won']:
        friendlyColor = 0x34ebb4
    elif not Match['teams'][team.lower()]['has_won'] and not Match['teams'][enemyTeam]['has_won']:
        friendlyColor = 0xd7d9db
    else:
        friendlyColor = 0xce0e0e

    if not Match['teams'][team.lower(
    )]['has_won'] and Match['teams'][enemyTeam]['has_won']:
        enemyColor = 0x34ebb4
    elif not Match['teams'][team.lower()]['has_won'] and not Match['teams'][enemyTeam]['has_won']:
        enemyColor = 0xd7d9db
    else:
        enemyColor = 0xce0e0e

    for i in players:
        if i['team'] == team:
            teamScore.append(i['stats']['score'])

    newScore = sorted(teamScore)
    print(newScore)
    await msg.edit(content='I have collected information for the ``friendly`` team. Now I\'m working on collecting information for the ``enemy`` team. Hang tight.')
    await asyncio.sleep(1)

    for ii in players:
        for num in range(5):
            if ii['stats']['score'] == newScore[num]:
                nameNum = f"{ii['name']}     {ii['stats']['kills']}/{ii['stats']['deaths']}/{ii['stats']['assists']}   {num}"
                teamSort.append(nameNum)
    print(teamSort)

    sortedTeam = sorted(teamSort, key=lastLetter)
    print(sortedTeam)

    finalMessage = ""

    for teamNum in range(4, -1, -1):
        finalMessage += f"\n{str(sortedTeam[teamNum])[:-1]} {newScore[teamNum]//metadata['rounds_played']}"

    Fembed = discord.Embed(
        title='Friendly Team',
        description=f'```{finalMessage}```',
        color=friendlyColor)
    await ctx.send(embed=Fembed)

    for i in players:
        if not i['team'] == team:
            enemyScore.append(i['stats']['score'])

    newEScore = sorted(enemyScore)
    print(newEScore)
    await msg.edit(content='I have collected information for the ``enemy`` team.')
    await asyncio.sleep(1)

    for ii in players:
        for num in range(5):
            if ii['stats']['score'] == newEScore[num]:
                nameNumE = f"{ii['name']}     {ii['stats']['kills']}/{ii['stats']['deaths']}/{ii['stats']['assists']}   {num}"
                enemySort.append(nameNumE)
    print(enemySort)

    sortedETeam = sorted(enemySort, key=lastLetter)

    finalEMessage = ""

    for teamNum in range(4, -1, -1):
        finalEMessage += f"\n{str(sortedETeam[teamNum])[:-1]} {round(newEScore[teamNum]/metadata['rounds_played'])}"

    Eembed = discord.Embed(
        title='Enemy Team',
        description=f'```{finalEMessage}```',
        color=enemyColor)
    await ctx.send(embed=Eembed)
    await msg.edit(content=f'Successfully got the leaderboard for match ``{lbnum + 1}`` of ``{name}#{tag}``.')

if __name__ == '__main__':
    leaderBoard(ctx, arg)


valm.py
 try:
    msg = await ctx.send(
              f"Your message has been registered and I'm working hard to query, read, format, and send you the last match for ``{arg}``. Hang tight!"
          )

          def mapImg(imap):
              if imap == "Haven":
                  return "https://static.wikia.nocookie.net/valorant/images/7/70/Loading_Screen_Haven.png/revision/latest/scale-to-width-down/1000?cb=20200620202335"
              elif imap == "Ascent":
                  return "https://static.wikia.nocookie.net/valorant/images/e/e7/Loading_Screen_Ascent.png/revision/latest/scale-to-width-down/1000?cb=20200607180020"
              elif imap == "Icebox":
                  return "https://static.wikia.nocookie.net/valorant/images/3/34/Loading_Icebox.png/revision/latest/scale-to-width-down/1000?cb=20201015084446"
              elif imap == "Split":
                  return "https://static.wikia.nocookie.net/valorant/images/d/d6/Loading_Screen_Split.png/revision/latest/scale-to-width-down/1000?cb=20200620202349"
              elif imap == "Bind":
                  return "https://static.wikia.nocookie.net/valorant/images/2/23/Loading_Screen_Bind.png/revision/latest/scale-to-width-down/1000?cb=20200620202316"
              elif imap == "Breeze":
                  return "https://www.ginx.tv/uploads2/Valorant/breeze_mapp.png"

          newArg = arg.split("#")
          name = newArg[0]
          tag = newArg[1]

          async with ctx.typing():
              response = requests.get(
                  f"https://api.henrikdev.xyz/valorant/v3/matches/na/{name}/{tag}"
              )
              jsonR = response.json()

              players = jsonR["data"]["matches"][0]["players"]["all_players"]
              convertD = datetime.fromtimestamp(
                  int(jsonR["data"]["matches"][0]["metadata"]["game_start"]) / 1000.0
              )
              finalT = convertD.strftime("%a, %b %d, %Y | %H:%M")

              def mode():
                  try:
                      return jsonR["data"]["matches"][0]["metadata"]["mode"]
                  except BaseException:
                      if KeyError:
                          return "Unknown"

              def map():
                  try:
                      return mapImg(
                          str(jsonR["data"]["matches"][0]["metadata"]["map"]))
                  except KeyError:
                      return "https://upload.wikimedia.org/wikipedia/commons/thumb/4/46/Question_mark_%28black%29.svg/200px-Question_mark_%28black%29.svg.png"

              for i in players:
                  if i["name"] == name or i["name"] == name.title():
                      team = str(i["team"]).lower()
                      puuid = i["puuid"]
                      if str(mode()) == "Deathmatch":
                          if (jsonR["data"]["matches"][0]["rounds"]
                                  [0]["winning_team"] == puuid):
                              color = 0x10B402
                          elif (
                              not jsonR["data"]["matches"][0]["rounds"][0]["winning_team"]
                              == puuid
                          ):
                              color = 0xDF0606
                          else:
                              color = 0x3B3D3C
                          won = i["stats"]["kills"]
                          lost = i["stats"]["deaths"]

                      else:
                          if (
                              jsonR["data"]["matches"][0]["teams"][team]["rounds_won"]
                              > jsonR["data"]["matches"][0]["teams"][team]["rounds_lost"]
                          ):
                              color = 0x10B402
                          elif (
                              jsonR["data"]["matches"][0]["teams"][team]["rounds_won"]
                              < jsonR["data"]["matches"][0]["teams"][team]["rounds_lost"]
                          ):
                              color = 0xDF0606
                          else:
                              color = 0xD1D1D1
                          won = jsonR["data"]["matches"][0]["teams"][team]["rounds_won"]
                          lost = jsonR["data"]["matches"][0]["teams"][team]["rounds_lost"]

                      global iconFile
                      iconFile = discord.File(
                          f"./imgs/agents/{i['character']}_icon.png")
                      global embedM
                      embedM = discord.Embed(
                          title=f"{i['name']}#{i['tag']}'s last match:",
                          description=f"**{mode()}** | ***{won}-{lost}***",
                          color=color,
                      )
                      embedM.add_field(name="Character: ", value=i["character"])
                      embedM.add_field(
                          name="KDA: ",
                          value=str(i["stats"]["kills"])
                          + "/"
                          + str(i["stats"]["deaths"])
                          + "/"
                          + str(i["stats"]["assists"]),
                      )
                      embedM.add_field(
                          name="Combat Score: ",
                          value=int(i["stats"]["score"])
                          // int(
                              jsonR["data"]["matches"][0]["metadata"]["rounds_played"]
                          ),
                          inline=True,
                      )
                      embedM.add_field(name="Date:", value=finalT)
                      embedM.add_field(
                          name="Duration:",
                          value=f"{int((jsonR['data']['matches'][0]['metadata']['game_length'])/1000)//60} minutes",
                      )
                      embedM.set_thumbnail(
                          url=f"attachment://{i['character']}_icon.png")
                      embedM.set_image(url=str(map()))
                      if mode() == "Unknown":
                          embedM.set_footer(
                              text="Wondering what happened to mode/map? Run command ``=error 1``.")
                      else:
                          embedM.set_footer(
                              text="https://github.com/5late/Pepe-Bot")
                      await asyncio.sleep(1)
                      await ctx.send(file=iconFile, embed=embedM)
                      await msg.edit(
                          content=f":smile: I successfully got last game stats for {name}#{tag}!"
                      )
    except BaseException:
      await msg.edit(content="Error 2||Error 404 :(")
      await ctx.send("Use command ``=error 2`` to see more information.")
